### DISCUSSION TITLE
how to get a FunctionSpace with a zero mean value in  global domain in firedrake?

### DISCUSSION THREAD

**Discussion by tu1620808124** [firedrakeproject/firedrake#2535] [Firedrake support]
I have seen a problem(sometimes Poisson equation) with a condition having zero mean value in the domain just like:

 ![1661308890480](https://user-images.githubusercontent.com/49664375/186307355-ca890c2e-3612-492f-9e37-847b4c20ea4f.png)
I want to know if there's a way to get this function space in firedrake, more detail can be seen in the reference:
>Masud A, Hughes T J R. A stabilized mixed finite element method for Darcy flow[J]. Computer methods in applied mechanics and engineering, 2002, 191(39-40): 4341-4370.

**Comment by dham** [comment]
You can't make this function space (or its subspaces) directly in Firedrake, but there are two ways of working with constrained function spaces when you actually need to solve problems. The first is to introduce a Lagrange multiplier imposing the zero integral constraint. This is documented [here](http://firedrakeproject.org/r-space.html). This directly imposes the constraint, but at the cost of making your variational problem harder to solve. The approach that is usually preferred is to specify your solution in the unconstrained space and to tell the solver about the resulting nullspace in the differential operator. This is covered [here](https://www.firedrakeproject.org/solving-interface.html#solving-singular-systems).
